### ISSUE TITLE
Bug in tiatoolbox/examples/inference-pipelines/idars.ipynb

### ISSUE THREAD

**Issue by DavidBAEpstein** [TissueImageAnalytics/tiatoolbox#325]
* TIA Toolbox version: 1.0.1
* Python version: 3.7.12
* Operating System: Macos 11.6.4

### Conclusions before Evidence (see Lewis Carroll _Sentence first—verdict afterwards_)
1. `PatchPredictor` needs to be rewritten so that its member function `predict` cannot contradict whether GPU is in use or not.
2. The example notebooks have no unittests. But these notebooks are supposed to be the most used part of TIA Toolbox. It's not very nice if a person's first contact with our package is a string of crashes. However, I know nothing about tests and nothing about how to use GPUs. How difficult would it be to test the running of each notebook under different environments, in particular with and without GPU?
3. In any case, it would be wise for all the example notebooks to be tested manually without GPU, just to make sure. I (David) can do that.

### Description
*Describe what you were trying to get done.*
I was testing idars.ipynb without GPU
*Tell us what happened, what went wrong, and what you expected to happen.*
Fatal errors. I expected the jupyter notebook to run to conclusion, but slowly

### What I Did
First I corrected one fatal error, caused by ON_GPU being undefined. (The user had been instructed in a text cell to skip the cell defining ON_GPU if not using a GPU.) Then I ran it again, and there was a different fatal error. _**The notebook seems to have been tested only with GPU.**_ The code cell causing the error had 3 statements, so I split the cell into three, and realised that the code was not set up to work without GPU.

The lines
```
tumour_predictor = PatchPredictor(
    pretrained_model='resnet18-idars-tumour',
    batch_size=64,
    num_loader_workers=8)
```
expect a GPU, but the subsequent lines
```
tumour_output = tumour_predictor.predict(
    imgs=wsi_file_list,
    mode='wsi',
    return_probabilities=True,
    on_gpu=ON_GPU)
```
do not expect a GPU, because at this point, `ON_GPU == False`

*Paste the command(s) you ran and the output.* 
I kept pressing Command Return, which on a Mac runs the next cell.

*If there was a crash, please include the traceback here.*
```
RuntimeError                              Traceback (most recent call last)
~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/torch/utils/data/dataloader.py in _try_get_data(self, timeout)
   1010         try:
-> 1011             data = self._data_queue.get(timeout=timeout)
   1012             return (True, data)

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/multiprocessing/queues.py in get(self, block, timeout)
    103                     timeout = deadline - time.monotonic()
--> 104                     if not self._poll(timeout):
    105                         raise Empty

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/multiprocessing/connection.py in poll(self, timeout)
    256         self._check_readable()
--> 257         return self._poll(timeout)
    258 

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/multiprocessing/connection.py in _poll(self, timeout)
    413     def _poll(self, timeout):
--> 414         r = wait([self], timeout)
    415         return bool(r)

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/multiprocessing/connection.py in wait(object_list, timeout)
    920             while True:
--> 921                 ready = selector.select(timeout)
    922                 if ready:

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/selectors.py in select(self, timeout)
    414         try:
--> 415             fd_event_list = self._selector.poll(timeout)
    416         except InterruptedError:

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/torch/utils/data/_utils/signal_handling.py in handler(signum, frame)
     65         # Python can still get and update the process status successfully.
---> 66         _error_if_any_worker_fails()
     67         if previous_handler is not None:

RuntimeError: DataLoader worker (pid 91493) is killed by signal: Segmentation fault: 11. 

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/var/folders/f2/pyj3khtr8xj8vr0059_y9lvh0000gp/T/ipykernel_91358/1154328519.py in <module>
      3     mode='wsi',
      4     return_probabilities=True,
----> 5     on_gpu=ON_GPU)

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/tiatoolbox/models/engine/patch_predictor.py in predict(self, imgs, masks, labels, mode, return_probabilities, return_labels, on_gpu, ioconfig, patch_input_shape, stride_shape, resolution, units, merge_predictions, save_dir, save_output)
    565                     return_probabilities=return_probabilities,
    566                     return_coordinates=return_coordinates,
--> 567                     on_gpu=on_gpu,
    568                 )
    569                 output_model["label"] = img_label

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/tiatoolbox/models/engine/patch_predictor.py in _predict_engine(self, dataset, return_probabilities, return_labels, return_coordinates, on_gpu)
    318             "labels": [],
    319         }
--> 320         for _, batch_data in enumerate(dataloader):
    321 
    322             batch_output_probabilities = self.model.infer_batch(

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/torch/utils/data/dataloader.py in __next__(self)
    528             if self._sampler_iter is None:
    529                 self._reset()
--> 530             data = self._next_data()
    531             self._num_yielded += 1
    532             if self._dataset_kind == _DatasetKind.Iterable and \

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/torch/utils/data/dataloader.py in _next_data(self)
   1205 
   1206             assert not self._shutdown and self._tasks_outstanding > 0
-> 1207             idx, data = self._get_data()
   1208             self._tasks_outstanding -= 1
   1209             if self._dataset_kind == _DatasetKind.Iterable:

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/torch/utils/data/dataloader.py in _get_data(self)
   1171         else:
   1172             while True:
-> 1173                 success, data = self._try_get_data()
   1174                 if success:
   1175                     return data

~/opt/miniconda3/envs/tiatoolbox/lib/python3.7/site-packages/torch/utils/data/dataloader.py in _try_get_data(self, timeout)
   1022             if len(failed_workers) > 0:
   1023                 pids_str = ', '.join(str(w.pid) for w in failed_workers)
-> 1024                 raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
   1025             if isinstance(e, queue.Empty):
   1026                 return (False, None)

RuntimeError: DataLoader worker (pid(s) 91493) exited unexpectedly

ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
ERROR: Unexpected segmentation fault encountered in worker.
```

**Comment by DavidBAEpstein** [comment]
Srijay and Shan: here are some more comments from me
[Issue325.txt](https://github.com/TissueImageAnalytics/tiatoolbox/files/8403466/Issue325.txt)

**Comment by DavidBAEpstein** [comment]
Fixed in PR #331, but, at this date, this PR has not yet been merged
